mathlibtools/file_status.py

```diff
@@ -1,53 +1,66 @@
-from dataclasses import dataclass, field
-from typing import Optional, Set
+import re
+from dataclasses import asdict, dataclass
+from typing import Any, Dict, Optional, Union
 
-@dataclass(frozen=True)
+import requests
+import yaml
+
+
+@dataclass()
 class FileStatus:
-  """Capture the file status descriptions in the wiki yaml.
-
-  `string_match` are the substrings that must be found, lowercase match
-  `color` is how the node should be colored if it has the status
-  """
-
-  color: str
-  prefix: Optional[str] = None
-  string_match: Set[str] = field(default_factory=set)
-  # colors from X11
-
-  @classmethod
-  def yes(cls) -> "FileStatus":
-    return cls("green", "Yes")
-
-  @classmethod
-  def pr(cls) -> "FileStatus":
-    return cls("lightskyblue", "No", {"pr"})
-
-  @classmethod
-  def wip(cls) -> "FileStatus":
-    return cls("lightskyblue", "No", {"wip"})
-
-  @classmethod
-  def no(cls) -> "FileStatus":
-    return cls("orange", "No")
-
-  # @classmethod
-  # def missing(cls) -> "FileStatus":
-  #   return cls("orchid1")
-
-  @classmethod
-  def ready(cls) -> "FileStatus":
-    return cls("turquoise1")
-
-  def matches(self, comment: str) -> bool:
-    return (self.prefix is None or comment.startswith(self.prefix)) and \
-        all(substring.lower() in comment.lower() for substring in self.string_match)
-
-  @classmethod
-  def assign(cls, comment: str) -> Optional["FileStatus"]:
-    for status in [cls.yes(), cls.pr(), cls.wip()]:
-      if status.matches(comment):
-        return status
-    # "No" but other comments
-    if cls.no().matches(comment) and len(comment) > 2:
-      return cls.no()
-    return None
+
+    ported: bool = False
+    mathlib4_pr: Optional[int] = None
+    mathlib3_hash: Optional[str] = None
+    comments: Optional[str] = None
+
+    @classmethod
+    def parse_old(cls, message: str) -> "FileStatus":
+        ported = False
+        mathlib4_pr: Optional[int] = None
+        mathlib3_hash: Optional[str] = None
+        if message.startswith("Yes"):
+            ported = True
+            if len(message.split()) > 2:
+                mathlib3_hash = message.split()[2]
+        if "mathlib4#" in message:
+            mathlib4_pr = int(re.findall(r"[0-9]+", message.replace("mathlib4#", ""))[0])
+        return cls(
+            ported=ported,
+            mathlib4_pr=mathlib4_pr,
+            mathlib3_hash=mathlib3_hash,
+        )
+
+    def asdict(self) -> Dict[str, Any]:
+        return {k: v for k, v in asdict(self).items() if v is not None}
+
+    def color(self) -> Optional[str]:
+        if self.ported:
+            return "green"
+        if self.mathlib4_pr:
+            return "lightskyblue"
+        if self.comments and "ready" in self.comments:
+            return "turquoise1"
+        return None
+
+
+@dataclass
+class PortStatus:
```

Comment:
Can you add a single API function that is do everyhing at once? For CI, I want a single function to call that is "load the portstatus from the wiki" so that I can process it. If there are multiple functions I'm more likely to be broken by future updates.

Reply:
`deserialize_old` will be that API for now. I made it have the input yaml as optional, so it does what you expect by default